src/client/kotlin/tech/thatgravyboat/skyblockapi/api/profile/friends/FriendsAPI.kt

```diff
@@ -0,0 +1,241 @@
+import tech.thatgravyboat.skyblockapi.utils.extentions.toIntValue
+import tech.thatgravyboat.skyblockapi.utils.regex.CommonRegexes
+import tech.thatgravyboat.skyblockapi.utils.regex.RegexGroup
+import tech.thatgravyboat.skyblockapi.utils.regex.RegexUtils.find
+import tech.thatgravyboat.skyblockapi.utils.regex.RegexUtils.findThenNull
+import tech.thatgravyboat.skyblockapi.utils.regex.component.find
+import tech.thatgravyboat.skyblockapi.utils.regex.component.toComponentRegex
+import tech.thatgravyboat.skyblockapi.utils.text.Text
+import tech.thatgravyboat.skyblockapi.utils.text.Text.send
+import tech.thatgravyboat.skyblockapi.utils.text.TextColor
+import tech.thatgravyboat.skyblockapi.utils.text.TextProperties.stripped
+import tech.thatgravyboat.skyblockapi.utils.text.TextStyle.bold
+import tech.thatgravyboat.skyblockapi.utils.text.TextStyle.color
+import tech.thatgravyboat.skyblockapi.utils.text.TextUtils.splitLines
+import java.util.*
+
+@Suppress("unused")
+@Module
+object FriendsAPI {
+
+    //region Regex
+    private val regexGroup = RegexGroup.CHAT.group("friend")
+    private val listGroup = regexGroup.group("list")
+
+    private val addedFriendRegex = regexGroup.create(
+        "add",
+        "^You are now friends with (?:\\[.+] )?(?<name>[a-zA-Z0-9_]+)"
+    )
+    private val removedFriendRegex = regexGroup.create(
+        "remove",
+        "^You removed (?:\\[.+] )?(?<name>[a-zA-Z0-9_]+) from your friends list!"
+    )
+
+    private val addedBestFriendRegex = regexGroup.create(
+        "bestfriend.add",
+        "^(?:\\[.+] )?(?<name>[a-zA-Z0-9_]+) is now a best friend!"
+    )
+    private val removeBestFriendRegex = regexGroup.create(
+        "bestfriend.remove",
+        "^(?:\\[.+] )?(?<name>[a-zA-Z0-9_]+) is no longer a best friend!"
+    )
+
+    private val pageFriendsListRegex = listGroup.create(
+        "page",
+        "^\\s*(?:<<)? Friends \\(Page (?<current>\\d+) of (?<max>\\d+)\\)"
+    )
+    private val friendEntryListRegex = listGroup.create(
+        "entry",
+        "^(?<name>\\S+) is "
+    ).toComponentRegex()
+
+    private val friendJoinLeaveRegex = regexGroup.create(
+        "joinleave",
+        "^Friend > (?<name>\\S+) (?<action>joined|left)"
+    )
+    //endregion
+
+    val friends: List<Friend>
+        get() = FriendStorage.friends
+
+    fun isFriend(name: String): Boolean = FriendStorage.getFriend(name) != null
+
+    fun isFriend(uuid: UUID): Boolean = FriendStorage.getFriend(uuid) != null
+
+    fun isBestFriend(name: String): Boolean = FriendStorage.getFriend(name)?.bestFriend ?: false
+
+    fun isBestFriend(uuid: UUID): Boolean = FriendStorage.getFriend(uuid)?.bestFriend ?: false
+
+    // Dealing with friends list
+    private var currentPage: Int = 0
+    private var maxPage: Int = 0
+    private var isInFriendsList = false
+    private val foundFriends: MutableSet<String> = mutableSetOf()
+
+    @Subscription
+    fun onChat(event: ChatReceivedEvent) {
```

Comment:
you want to receive cancelled events and be the very first to receive the event too